### PR TITLE
Tweak dynamic supervisor with inline counter

### PR DESCRIPTION
**Pull request by batate**
What about this? I inlined a minimal counter. It's easy to tell what it does, and it doesn't get in the way of the dynamic supervisor description.

### CODE REVIEW

File: lib/elixir/lib/dynamic_supervisor.ex [hunk 1]
```diff
@@ -22,17 +22,29 @@ defmodule DynamicSupervisor do
 
   The options given in the child specification are documented in `start_link/1`.
 
-  Once the dynamic supervisor is running, we can start children
-  with `start_child/2`, which receives a child specification:
+  Once the dynamic supervisor is running, we can use it to start children
+  on demand. Given this `Counter` GenServer:
```
Comment (josevalim) [inline lib/elixir/lib/dynamic_supervisor.ex:26]:
```suggestion
  on demand. Given this sample `GenServer`:
```

let's highlight the GenServer bits.

File: lib/elixir/lib/dynamic_supervisor.ex [hunk 2]
```diff
@@ -22,17 +22,29 @@ defmodule DynamicSupervisor do
...
+        use GenServer
 
-      {:ok, agent2} = DynamicSupervisor.start_child(MyApp.DynamicSupervisor, {Agent, fn -> %{} end})
-      Agent.get(agent2, & &1)
-      #=> %{}
+        def start_link(initial), do: GenServer.start_link(__MODULE__, initial)
+        def inc(pid), do: GenServer.call(pid, :inc)
+
+        def init(initial), do: {:ok, initial}
+        def handle_call(:inc, _, count), do: {:reply, count, count + 1}
```
Comment (josevalim) [inline lib/elixir/lib/dynamic_supervisor.ex:35]: Looks good to me! I believe you used the single line definitions for conciseness but I would go with the multi-line one for clarify.

**Review by batate — COMMENTED** [review]
Agree on all counts.

**Review by whatyouhide — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**